src/Text/SlamSearch/Parser/Term.purs

```diff
@@ -0,0 +1,202 @@
+module Text.SlamSearch.Parser.Terms where
+
+import Text.SlamSearch.Parser.Values
+import Text.SlamSearch.Parser.Tokens
+import Text.SlamSearch.Parser.Utils
+
+import Text.Parsing.Parser
+import Text.Parsing.Parser.Combinators
+import Control.Apply
+import Control.Alt
+import Control.Alternative
+import Data.Tuple
+import Data.Either
+
+data SimplePredicate =
+  ContainsP Value
+  | EqP Value
+  | GtP Value
+  | GteP Value
+  | LteP Value
+  | LtP Value
+  | NeP Value
+  | LikeP Value
+
+instance showSimplePredicate :: Show SimplePredicate where
+  show a = case a of
+    ContainsP p -> "ContainsP(" <> show p <> ")"
+    EqP p -> "EqP(" <> show p <> ")"
+    GtP p -> "GtP(" <> show p <> ")"
+    GteP p -> "GteP(" <> show p <> ")"
+    LteP p -> "GteP(" <> show p <> ")"
+    LtP p -> "LtP(" <> show p <> ")"
+    NeP p -> "NeP(" <> show p <> ")"
+    LikeP p -> "LikeP(" <> show p <> ")"
+
+
+instance eqSimplePredicate :: Eq SimplePredicate where
+  (==) (ContainsP p) (ContainsP p') = p == p'
+  (==) (EqP p) (EqP p') = p == p'
+  (==) (GtP p) (GtP p') = p == p'
+  (==) (GteP p) (GteP p') = p == p'
+  (==) (LteP p) (LteP p') = p == p'
+  (==) (LtP p) (LtP p') = p == p'
+  (==) (NeP p) (NeP p') = p == p'
+  (==) (LikeP p) (LikeP p') = p == p'
+  (==) _ _ = false
+  (/=) a b = not $ a == b
+
+data Label =
+  Common String
+  | Meta String
+
+instance showLabel :: Show Label where
+  show a = case a of
+    Common p -> "Common(" <> show p <> ")"
+    Meta p -> "Meta(" <> show p <> ")"
+
+instance eqLabel :: Eq Label where
+  (==) (Common p) (Common p') = p == p'
+  (==) (Meta p) (Meta p') = p == p'
+  (==) _ _ = false
+  (/=) a b = not $ a == b
+
+data SearchTermSimple =
+  SimplePredicate SimplePredicate
+  | InfieldPredicate [Label] SimplePredicate
```

Comment:
If the array is empty, this reduces to the other case, so you could either (a) simplify it to just `SearchTermSimple [Label] SimplePredicate` (which would represent all cases), or force the array to be non-empty by using either `Cofree Option Label` or `Tuple Label [Label]`.